addOns/scripts/src/main/java/org/zaproxy/zap/extension/scripts/scanrules/ActiveScriptSynchronizer.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2024 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.scripts.scanrules;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.core.scanner.PluginFactory;
+import org.zaproxy.addon.commonlib.scanrules.ScanRuleMetadata;
+import org.zaproxy.addon.commonlib.scanrules.ScanRuleMetadataProvider;
+import org.zaproxy.zap.extension.script.ExtensionScript;
+import org.zaproxy.zap.extension.script.ScriptWrapper;
+
+public class ActiveScriptSynchronizer {
+
+    private static final Logger LOGGER = LogManager.getLogger(ActiveScriptSynchronizer.class);
+
+    private ExtensionScript extScript;
+
+    private final Map<Integer, ActiveScriptScanRule> scriptIdToScanRuleMap = new HashMap<>();
+
+    public void scriptAdded(ScriptWrapper script) {
+        try {
+            var metadata = getMetadataForScript(script);
+            if (metadata == null) {
+                return;
+            }
+            // TODO: Should we check non-script scan rules as well?
```

Comment:
Per my note above yes I think we should, or at least document the behaviour if there is a collision.